mount/mount.go

```diff
@@ -22,3 +47,48 @@ func All(mounts []Mount, target string) error {
 	}
 	return nil
 }
+
+// TempMount mounts the provided mounts to a temp dir, and pass the temp dir to f.
+// The mounts are valid during the call to the f.
+// Finally we will unmount and remove the temp dir regardless of the result of f.
+func TempMount(mounts []Mount, m Mounter, f func(root string) error) error {
+	if len(mounts) == 0 {
+		return errdefs.ErrInvalidArgument
+	}
+	if m == nil {
+		m = mounter{}
+	}
+
+	root, err := ioutil.TempDir("", "containerd-TempMount")
+	if err != nil {
+		return errors.Wrapf(err, "failed to create %s", root)
+	}
+	// We use Remove here instead of RemoveAll.
+	// The RemoveAll will delete the temp dir and all children it contains.
+	// When the Unmount fails, if we use RemoveAll, We will incorrectly delete data from mounted dir.
+	// if we use Remove,even though we won't successfully delete the temp dir,
+	// but we only leak a temp dir, we don't loss data from mounted dir.
+	// For details, please refer to #1868 #1785.
+	defer func() {
+		if err := os.Remove(root); err != nil {
+			log.L.Errorf("Failed to remove the temp dir %s: %v", root, err)
+		}
+	}()
+
+	// We should do defer first, or we will not do Unmount when only partful Mount fail.
+	defer func() {
+		if uerr := m.Unmount(root); uerr != nil {
+			uerr = errors.Wrapf(uerr, "failed to unmount %s", root)
+			if err == nil {
+				err = uerr
+			} else {
+				err = errors.Wrap(uerr, err.Error())
```

Comment:
I'm not sure that wrapping the error is correct in this case, because it's not handling a returned error, but erroring again during a cleanup.

Maybe this should instead just log `uerr` (and mention that there was a previous failure) and keep `err` as the return?

```go
log.L.Errorf("failed to unmount after mount failed: %s", uerr)
```

Reply:
I have added the named return error, so we don't need to log it. I think It is up to the user.